oauth2/jwt.go

```diff
@@ -30,56 +42,121 @@ func NewJWT(secret string, jwksurl string) *JWT {
+	token, err := jwt.Parse([]byte(src), options...)
+	if err != nil {
+		// Hack to make the rror messages compatible with previous incarnation
+		switch {
+		case errors.Is(err, jwt.ErrInvalidIssuedAt()):
+			return nil, fmt.Errorf(`Token used before issued`)
+		case errors.Is(err, jwt.ErrTokenExpired()):
+			return nil, fmt.Errorf(`Token is expired`)
+		default:
+		}
+		return nil, err
 	}
-
-	return nil
+	return token, nil
 }
 
 // ValidPrincipal checks if the jwtToken is signed correctly and validates with Claims.  lifespan is the
 // maximum valid lifetime of a token.  If the lifespan is 0 then the auth lifespan duration is not checked.
 func (j *JWT) ValidPrincipal(ctx context.Context, jwtToken Token, lifespan time.Duration) (Principal, error) {
-	gojwt.TimeFunc = j.Now
 
-	// Check for expected signing method.
-	alg := j.KeyFunc
+	var options = []jwt.ParseOption{
+		jwt.WithKeyProvider(j),
+		jwt.WithRequiredClaim(jwt.SubjectKey),
+		jwt.WithClock(jwt.ClockFunc(j.Now)),
+	}
+
+	// If the duration of the claim is longer than the auth lifespan then this is
+	// an invalid claim because server assumes that lifespan is the maximum possible
+	// duration.  However, a lifespan of zero means that the duration comparison
+	// against the auth duration is not needed.
+	if lifespan > 0 {
+		options = append(options, jwt.WithValidator(jwt.ValidatorFunc(func(_ context.Context, tok jwt.Token) jwt.ValidationError {
+			exp := tok.Expiration().Round(time.Second)
+			iat := tok.IssuedAt().Round(time.Second)
+
+			// If the duration of the claim is longer than the auth lifespan then this is
+			// an invalid claim because server assumes that lifespan is the maximum possible
+			// duration.  However, a lifespan of zero means that the duration comparison
+			// against the auth duration is not needed.
+			if lifespan > 0 && exp.Sub(iat) > lifespan {
+				return jwt.NewValidationError(fmt.Errorf("claims duration is different from auth lifespan"))
+			}
+			return nil
+		})))
+	}
+
+	token, err := parseToken(string(jwtToken), options...)
+	if err != nil {
+		return Principal{}, err
+	}
+
+	var org, grp string
+	if v, ok := token.Get(`org`); ok {
+		if s, ok := v.(string); ok {
+			org = s
+		}
+	}
+	if v, ok := token.Get(`grp`); ok {
+		if s, ok := v.(string); ok {
+			grp = s
+		}
+	}
 
-	return j.ValidClaims(jwtToken, lifespan, alg)
+	return Principal{
+		Subject:      token.Subject(),
+		Issuer:       token.Issuer(),
+		Organization: org,
+		Group:        grp,
+		ExpiresAt:    token.Expiration(),
+		IssuedAt:     token.IssuedAt(),
+	}, nil
 }
 
-// KeyFunc verifies HMAC or RSA/RS256 signatures
-func (j *JWT) KeyFunc(token *gojwt.Token) (interface{}, error) {
-	if _, ok := token.Method.(*gojwt.SigningMethodHMAC); ok {
-		return []byte(j.Secret), nil
-	} else if _, ok := token.Method.(*gojwt.SigningMethodRSA); ok {
-		return j.KeyFuncRS256(token)
+func (j *JWT) initCache(ctx context.Context) {
+	cache := jwk.NewCache(ctx)
```

Comment:
It seems to me that the cache is reloaded with this context remembered, it in fact has to rely upon background/TODO context

Reply:
changed to content.TODO() with a comment